subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/DefaultFileCollectionSnapshotter.java

```diff
@@ -47,19 +47,46 @@ public DefaultFileCollectionSnapshotter(FileSystemAccess fileSystemAccess, Gener
 
     @Override
     public FileSystemSnapshot snapshot(FileCollection fileCollection) {
+        return snapshotResult(fileCollection).getSnapshot();
+    }
+
+    @Override
+    public Result snapshotResult(FileCollection fileCollection) {
         SnapshottingVisitor visitor = new SnapshottingVisitor();
         ((FileCollectionInternal) fileCollection).visitStructure(visitor);
-        return CompositeFileSystemSnapshot.of(visitor.getRoots());
+        return new DefaultResult(CompositeFileSystemSnapshot.of(visitor.getRoots()), visitor.isOnlyFileTrees());
+    }
+
+    private static class DefaultResult implements Result {
+        private final FileSystemSnapshot fileSystemSnapshot;
+        private final boolean isFileTree;
+
+        public DefaultResult(FileSystemSnapshot fileSystemSnapshot, boolean isFileTree) {
+            this.fileSystemSnapshot = fileSystemSnapshot;
+            this.isFileTree = isFileTree;
+        }
+
+        @Override
+        public FileSystemSnapshot getSnapshot() {
+            return fileSystemSnapshot;
+        }
+
+        @Override
+        public boolean isTree() {
+            return isFileTree;
+        }
     }
 
     private class SnapshottingVisitor implements FileCollectionStructureVisitor {
         private final List<FileSystemSnapshot> roots = new ArrayList<>();
+        private boolean isOnlyFileTrees = true;
 
         @Override
         public void visitCollection(FileCollectionInternal.Source source, Iterable<File> contents) {
             for (File file : contents) {
                 fileSystemAccess.read(file.getAbsolutePath(), roots::add);
             }
+            isOnlyFileTrees = false;
```

Comment:
What is `GenericFileTree` again? In what cases do we end up with one?

Reply:
We only end up in `visitGenericFileTree`, if the file collection is an archive file tree without a backing file. To my knowledge, this can only happen with tar trees from resources. I created a test here: https://github.com/gradle/gradle/blob/984582a728c91bfbdac30e22a4f1389396c2c490/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/SkipWhenEmptyIntegrationTest.groovy#L80

Reply:
Let's handle this as part of https://github.com/gradle/gradle/issues/18854.